src/files/infrastructure/FileJSDataverseRepository.ts

```diff
@@ -284,12 +287,51 @@ export class FileJSDataverseRepository implements FileRepository {
   }
 
   uploadFile(
-    _datasetId: number | string,
-    _file: FileHolder,
-    _progress: (now: number) => void,
-    _abortController: AbortController
+    datasetId: number | string,
+    file: FileHolder,
+    progress: (now: number) => void,
+    abortController: AbortController,
+    storageIdSetter: (storageId: string) => void
   ): Promise<void> {
-    // TODO:
-    return new Promise(() => {})
+    return jsUploadFile
+      .execute(datasetId, file.file, progress, abortController)
+      .then(storageIdSetter)
+      .catch((error: ReadError) => {
+        throw new Error(error.message)
+      })
+  }
+
+  addUploadedFile(datasetId: number | string, files: FileUploadState[]): Promise<void[]> {
+    const all: Promise<void>[] = []
+    files.forEach((file) => {
+      const f: FileHolder = {
+        file: {
+          lastModified: file.fileLastModified,
+          name: file.fileName,
+          webkitRelativePath: file.fileDir ? file.fileDir + '/' + file.fileName : file.fileName,
+          size: file.fileSize,
+          type: file.fileType,
+          arrayBuffer: function (): Promise<ArrayBuffer> {
```

Comment:
I think mocking all of this native methods from File interface could bring problems (arrayBuffer, slice, stream, text).
If the use case is expecting an Original File you should store the original file somewhere I believe inside FileUploadState. Maybe an `original` or `nativeFile` property and then you can use it here. 
So you don't mix original File properties with properties you are modifying or using in the state.

Reply:
It is due to a bad design of the current js-dataverse call. If it was implement literately, as the documentation states:

"Once the files exists in the s3 bucket, a final API call is needed to add all the files to the Dataset. In this API call, additional metadata is added using the “jsonData” parameter. jsonData for this call is an array of objects that normally include information such as a file description, tags, provenance, whether the file is restricted, etc. For direct uploads, the jsonData object must also include values for:..."

We would not be having this conversation and some of the other remarks. That last call does NOT need files! It needs the description, tags, restriction boolean, etc. Exactly what the UI part does: provide editable, adjusted by the user metadata to finalize the addition of the files to the dataset.

Reply:
https://guides.dataverse.org/en/latest/developers/s3-direct-upload-api.html#direct-add-to-dataset-api

Reply:
Ok so it is just a workaround  I understand, thanks for the clarification 👍🏼 
Please indicate in the description of the PR which parts of the code ( very roughly ) have some workaround to make it work, so we will know later what needs to be refactored.

Reply:
I extended the description of that point in the follow-up issue: https://github.com/IQSS/dataverse-frontend/issues/431